### PR TITLE
feat(binder): Bind `natural join`

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Bind `natural join`

Questions:
1. Should natural join also bind the `_row_id` column? I've chosen to manually exclude any column with such a name.

Depends on the lateral_context defined in https://github.com/singularity-data/risingwave/pull/3788
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
Closes: https://github.com/singularity-data/risingwave/issues/1633

### CODE REVIEW

File: src/frontend/src/binder/relation/join.rs [hunk 1]
```diff
@@ -107,7 +111,52 @@ impl Binder {
         Ok(match constraint {
             JoinConstraint::None => (ExprImpl::literal_bool(true), None),
             JoinConstraint::Natural => {
-                return Err(ErrorCode::NotImplemented("Natural join".into(), 1633.into()).into())
+                // First, we identify columns with the same name.
+                let old_context = self.context.clone();
+                self.push_lateral_context();
+                // Bind this table factor to an empty context
+                let table_factor = table_factor.unwrap();
+                let right_table = get_table_name(&table_factor);
```
Comment (xiangjinwu) [inline src/frontend/src/binder/relation/join.rs:119]: FYI using `get_table_name` on right_table_factor would not work once we support nested join.

For example, in `a natural join (b join c on true)`, we need to find the common columns between `a` and `(b join c on true)`. The right part of join does not even have a single table name, and the whole `get_table_name` was just a hack/workaround for `join-using`. `join-using` would also break after nested join.

**Comment by xiangjinwu** [comment]
Not necessarily handled in this PR, but in case you are interested for full support: there are other tricky parts of `natural join` and `join-using`, as they only output a single copy of the common column.

```
select b from ab natural join bc; -- reference to b does not raise ambiguous error

insert into ab values (1, 1), (2, 2);
insert into bc values (2, 2), (3, 3);
select *, b as unqualified, ab.b as left, bc.b as right from ab natural full join bc;

 b | a | c | unqualified | left | right 
---+---+---+-------------+------+-------
 1 | 1 |   |           1 |    1 |      
 2 | 2 | 2 |           2 |    2 |     2
 3 |   | 3 |           3 |      |     3
(3 rows)
```

In the output, `*` and unqualified `b` refers to a "pseudo" column with values `1, 2, 3`, while `ab.b` is `1, 2, null` and `bc.b` is `null, 2, 3`.

**Comment by jon-chuang** [comment]
Indeed, its more non-trivial to support natural join than expected. We will need executor-level support to be able to merge output columns to coalesce values from both left and right when nulls are present for non-inner joins.

File: src/frontend/src/binder/relation/join.rs (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/src/binder/relation/join.rs:115]: Actually, the more I think of it, we could recursively resolve the source table names in nested join.

**Comment by jon-chuang** [comment]
Hi @xiangjinwu , I've decided to go with the option of adding an `projection: Option<Vec<ExprImpl>>` field to `BoundJoin` and convert it to a `LogicalProject` if it is not none in planner. It produces the right results as can be seen from the planner tests.

File: src/frontend/src/binder/relation/join.rs (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/src/binder/relation/join.rs:116]: Anyway, this is not relevant to the PR, we will solve it in nested join PR

**Comment by jon-chuang** [comment]
Actually, I realize my approach does not really meet the requirement fully. In my approach, it is not possible to access the original columns as shown in the example. Achieving that is a little more challenging. 

We need to keep all columns, and decide which columns to project away based on whether the select clause references those columns. The unqualified column is coalesced or taken from left or right.

Basically, what we could do is mark unqualified columns that are involved in natural joins as such, and convert references to them to coalesce. 

Alternatively, we could convert all unqualified references to coalesce. But I think it would break assumptions, so we shouldn't do this. 

Furthermore, it appears that select * would also need to be handled specially. Or it may not, depending on the semantics. if what select * does is to reference all columns by name an look for them in the context, we are ok, because we can use the previously outlined codepath to solve it. But if we are simply looking at the context itself, we need to change it to looking for the columns by name...

Otherwise, we can simply use the binder context where columns are marked as being natural columns to disambiguate...

**Pull request closed without merging**